### PR TITLE
PKI Health Check Command

### PR DESCRIPTION
**Pull request by cipherboy**
This pull requests adds initial support for the `vault pki health-check` command, with two health checks. More will be added in future PRs, but I wanted to get this opened for review before we get too far and add too many tests.

---

TODO:

 - [x] Finish CRL HC implementation
 - [x] Add tests for initial health checks
 - [x] Add a little code reuse between PKI tests
 - [x] Finish return code parsing

### CODE REVIEW

File: command/healthcheck/common.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+func (e *Executor) FetchIfNotFetched(op logical.Operation, rawPath string) (*PathFetch, error) {
+	path := e.templatePath(rawPath)
+
+	byOp, present := e.Resources[path]
+	if present && byOp != nil {
+		result, present := byOp[op]
+		if present && result != nil {
+			return result, nil
```
Comment (cipherboy) [inline command/healthcheck/common.go:143]: Need to re-surface the error for consistency.

File: command/healthcheck/pki_crl_validity_period.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+			ret := Result{
+				Status:   ResultInsufficientPermissions,
+				Endpoint: "/{{mount}}/config/crl",
+				Message:  "This prevents the health check from seeing if the CRL is disabled and dropping the severity of this check accordingly.",
+			}
+
+			if e.Client.Token() == "" {
+				ret.Message = "No token available so unable read authenticated CRL configuration for this mount. " + ret.Message
+			} else {
+				ret.Message = "This token lacks so permission to read the CRL configuration for this mount. " + ret.Message
```
Comment (cipherboy) [inline command/healthcheck/pki_crl_validity_period.go:140]:
```suggestion
				ret.Message = "This token lacks permission to read the CRL configuration for this mount. " + ret.Message
```

File: command/healthcheck/common.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+	return p.IsOK() && p.SecretParseError == nil && p.Secret != nil
+}
+
+func (p *PathFetch) FetchSurfaceError() error {
+	if p.IsOK() || p.IsSecretPermissionsError() || p.IsUnsupportedPathError() || p.IsMissingResource() {
+		return nil
+	}
+
+	if strings.Contains(p.FetchError.Error(), "route entry not found") {
+		return fmt.Errorf("Error making API request: was a bad mount given?\n\nOperation: %v\nPath: %v\nOriginal Error:\n%v", p.Operation, p.Path, p.FetchError)
```
Comment (stevendpclark) [inline command/healthcheck/common.go:209]: nit: %w, I won't call out all the examples to flood the PR, but might want to have a look through as there are a bunch we could tweak.

```suggestion
		return fmt.Errorf("Error making API request: was a bad mount given?\n\nOperation: %v\nPath: %v\nOriginal Error:\n%w", p.Operation, p.Path, p.FetchError)
```

File: command/healthcheck/pki_crl_validity_period.go (hunk 2 above)
Comment (stevendpclark) [inline command/healthcheck/pki_crl_validity_period.go:140]: I accidentally deleted this comment from @cipherboy, adding it back.
 
```suggestion
				ret.Message = "This token lacks permission to read the CRL configuration for this mount. " + ret.Message
```

**Review by stevendpclark — APPROVED** [review]
A small nit with the %w, but everything looks good. I'm assuming doc updates will come with a later PR?

**Comment by cipherboy** [comment]
Thanks! Pushed an update to those comments and will auto-merge.

**Pull request merged**